### ISSUE TITLE
Using Poisson objective and a baseline gives negative predictions

### ISSUE THREAD

**Issue by BWallyn** [catboost/catboost#1323]
Problem: Using Poisson objective and a baseline gives negative predictions
catboost version: 0.23.2
Operating System: Linux, redhat enterprise 7
CPU: 40 * Intel Corporation Xeon E7 v3/Xeon E5 v3/Core i7
# GPU: No

Hello, 
I am trying to predict an integer target using Poisson objective. I need an offset so I'm using the set_baseline() function on the Pools. However, the prediction given by Catboost is very bad and when setting the baseline with a log gives me negative predictions. 
Am I missing something ? Or Could it be a problem similar to the issue #1184 but for the baseline ?

Thank you for your help !

**Comment by alanchalk** [comment]
Similar issue
catboost version 0.24.4
OS Windows 10

Hi,

I had a similar issue - that trying to do Poisson regression with an offset, gives very poor results.

I have no solution, but a workaround in my case seems to be to predict target / offset (not taking the log of the offset).

And then setting weights to be the offset.

In insurance claims frequency applications, where offset is exposure, these two models are equivalent.

I also noticed that GPU training in either case does not work for me.  But switching back to CPU gives reasonable results.

If anyone else comes here and would like to talk about a concrete case:

import random
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from catboost import CatBoostRegressor
from catboost import Pool as pool

# to make this notebook's output stable across runs
np.random.seed(2021)

# define some features
ages = [age for age in range(17, 100) for _ in range(100)]
gender = ["M"]*50 + ["F"]*50
gender = gender * (100 - 17)
df_all = pd.DataFrame({'age': ages,
                      'gender': gender})

# let frequency vary by feature
# set different frequencies for M / F
df_all["freq_M"] = ((70 - df_all['age']) /53)**2 
df_all["freq_F"] = (df_all['age'] - 16) / 99


df_all["freq"] = df_all["freq_M"]
idx_F = df_all["gender"] == "F"
df_all.loc[idx_F, "freq"] = df_all.loc[idx_F, "freq_F"]
del idx_F
df_summary = df_all.groupby(['age', 'gender']).aggregate({'freq': sum})
df_summary = df_summary.unstack(level=1)
df_summary.plot(kind='bar', subplots=True)

# Now add random exposure
df_all['ex'] = np.random.uniform(0, 1, df_all.shape[0])

# create expected frequency for each example
df_all['mu'] = df_all['ex'] * df_all['freq']

# simulate random number of claims based on poisson distribution
df_all['nucl'] = np.random.poisson(df_all['mu'])

# prepare catboost data
X = df_all[['age', 'gender']]
# create variables for the numerical and categorical column names:
cat_att = list(X.select_dtypes(include=['category', 'O']).columns)

# choose which formulation of the model you would like
# formulation 1: model number of claims and use log ex for offset
# This formulation does not work on catboost 0.24.4
y = df_all['nucl']
offset = np.log(df_all['ex'])
D_train = pool(X, y, cat_features=cat_att, baseline=offset)

# formulation 1: model number of claims and use log ex for offset
y = df_all['nucl'] / df_all['ex']
D_train = pool(X, y, cat_features=cat_att, weight = df_all['ex'])

# task type GPU does not give correct results
model    = CatBoostRegressor(task_type="CPU",
                             #task_type="GPU",
                             #objective='RMSE',
                             objective='Poisson',
                             random_seed=2020,
                             verbose=True,
                             allow_writing_files=False
                             )

model.fit(D_train)

# when modelling log offset - results are negative and are helped a little using exp
# though this is probably not what is going wrong
#df_all['m1'] = np.exp(model.predict(D_train))
#df_all['m2'] = df_all['m1'] * df_all['ex']


# when modelling frequency and using weights
df_all['m1'] = model.predict(D_train)
df_all['m2'] = df_all['m1'] * df_all['ex']

# final table shows results are good - ie overfitting and close to actual claims
df_all.groupby(['age', 'gender']).aggregate({'mu': sum,
                                             'nucl': sum,
                                            'm2': sum})

# using GPU or alternative model formulation gives poor results

**Comment by EvanZ** [comment]
I know this is very old, but looks like not many people are using Catboost for actuarial applications, perhaps, because exposure (offset) is not directly supported. I do not think it is equivalent to set sample weights as the offsets or to use, for example, claims/exposure, as the target since Poisson inherently is supposed to model discrete distributions. 

Is the only way to handle exposure "correctly" in Catboost for me to implement my own custom loss function?
